stan/math/rev/mat/functor/jacobian.hpp

```diff
@@ -18,20 +18,17 @@ void jacobian(const F& f, const Eigen::Matrix<double, Eigen::Dynamic, 1>& x,
   using Eigen::Matrix;
   start_nested();
   try {
-    Matrix<var, Dynamic, 1> x_var(x.size());
-    for (int k = 0; k < x.size(); ++k)
-      x_var(k) = x(k);
+    Matrix<var, Dynamic, 1> x_var(x);
     Matrix<var, Dynamic, 1> fx_var = f(x_var);
     fx.resize(fx_var.size());
-    for (int i = 0; i < fx_var.size(); ++i)
-      fx(i) = fx_var(i).val();
     J.resize(fx_var.size(), x.size());
-    for (int i = 0; i < fx_var.size(); ++i) {
-      if (i > 0)
-        set_zero_all_adjoints_nested();
+    fx = fx_var.val();
+    grad(fx_var(0).vi_);
+    J.row(0) = x_var.adj();
```

Comment:
Also this is done by rows, does anyone know how much we need to change to have this done by columns. I'm assuming this Eigen matrix is column major order so we'll get better speed from doing this by column

Reply:
That's just how Jacobians are defined---each row is a gradient.

For reverse mode, it might be better to pay a single transposition cost at the end.  That is, instead of building up an N x M matrix row major, build up an M x N matrix column major, then do a single transpose at the end.  Given that you only hit each element once, it might not make a difference on modest size Jacobians.

Once we have forward-mode thoroughly tested, it might be faster to calculate Jacobians using that, which would also be naturally column major.

Getting rid of the extra test and just going from 1 probably helps.

Reply:
Word! That makes sense. I added the transposition at the end and do things by col. Idt it would matter that much either besides large matrices tho'.